system-helper/flatpak-system-helper.c

```diff
@@ -236,44 +235,23 @@ handle_deploy (FlatpakSystemHelper   *object,
       return TRUE;
     }
 
-  is_update = (arg_flags & FLATPAK_HELPER_DEPLOY_FLAGS_UPDATE) != 0;
   no_deploy = (arg_flags & FLATPAK_HELPER_DEPLOY_FLAGS_NO_DEPLOY) != 0;
   local_pull = (arg_flags & FLATPAK_HELPER_DEPLOY_FLAGS_LOCAL_PULL) != 0;
   reinstall = (arg_flags & FLATPAK_HELPER_DEPLOY_FLAGS_REINSTALL) != 0;
 
   deploy_dir = flatpak_dir_get_if_deployed (system, arg_ref, NULL, NULL);
 
-  if (deploy_dir)
+  if (deploy_dir && !reinstall)
     {
       g_autofree char *real_origin = NULL;
-      if (!is_update)
+      real_origin = flatpak_dir_get_origin (system, arg_ref, NULL, NULL);
+      if (g_strcmp0 (real_origin, arg_origin) != 0)
         {
-          if (!reinstall)
-            {
-              /* Can't install already installed app */
-              g_dbus_method_invocation_return_error (invocation, FLATPAK_ERROR, FLATPAK_ERROR_ALREADY_INSTALLED,
-                                                     "%s is already installed", arg_ref);
-              return TRUE;
-            }
-        }
-      else
-        {
-          real_origin = flatpak_dir_get_origin (system, arg_ref, NULL, NULL);
-          if (g_strcmp0 (real_origin, arg_origin) != 0)
-            {
-              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_INVALID_ARGS,
-                                                     "Wrong origin %s for update", arg_origin);
-              return TRUE;
-            }
+          g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_INVALID_ARGS,
+                                                 "Wrong origin %s for update", arg_origin);
+          return TRUE;
         }
     }
-  else if (!deploy_dir && is_update)
-    {
-      /* Can't update not installed app */
-      g_dbus_method_invocation_return_error (invocation, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED,
-                                             "%s is not installed", arg_ref);
```

Comment:
similarly, this case here just doesn't happen anymore, where we had the caller claim that this is an update, but there is in fact no deploy